include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,385 @@
+
+    // Create new population of candidate from the present elite population
+    // Have P_t, generate G_t using P_t
+    BinaryTournamentSelection(population, lowerBound, upperBound);
+
+    // Evaluate the objectives for the new population
+    calculatedObjectives.resize(population.size());
+    EvaluateObjectives(population, objectives, calculatedObjectives);
+
+    // Perform fast non dominated sort on P_t ∪ G_t
+    ranks.resize(population.size());
+    FastNonDominatedSort<MatType>(fronts, ranks, calculatedObjectives);
+
+    // Perform crowding distance assignment
+    crowdingDistance.resize(population.size());
+
+    for (size_t fNum = 0; fNum < fronts.size(); fNum++)
+      CrowdingDistanceAssignment(fronts[fNum],
+                                 crowdingDistance,
+                                 lowerBound,
+                                 upperBound);
+
+    // Sort based on crowding distance
+    std::sort(population.begin(),
+              population.end(),
+              [this, ranks, crowdingDistance, population](MatType candidateP,
+                                                          MatType candidateQ)
+              {
+                size_t idxP, idxQ;
+                for (int i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i],
+                                         candidateP,
+                                         "absdiff",
+                                         epsilon))
+                    idxP = i;
+
+                  if (arma::approx_equal(population[i],
+                                         candidateQ,
+                                         "absdiff",
+                                         epsilon))
+                    idxQ = i;
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // Yield a new population P_{t+1} of size populationSize
+    population.resize(populationSize);
+  }
+
+  // Set the candidates from the best front as the output
+  std::vector<MatType> bestFront;
+
+  for (size_t f: fronts[0])
+    bestFront.push_back(population[f]);
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+//! Evaluate the objectives for the entire population.
+template<typename MatType,
+         typename ...ArbitraryFunctionType>
+inline void NSGA2::EvaluateObjectives(
+    std::vector<MatType>& population,
+    std::tuple<ArbitraryFunctionType...>& objectives,
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives)
+{
+  // TODO: The std::get<> needs compile time constant indices!
+  for (size_t i = 0; i < numObjectives; i++)
+    calculatedObjectives[i] = std::get<i>(objectives).Evaluate(population[i]);
```

Comment:
This fails as `std::get<i>` [requires](https://thispointer.com/c11-stdtuple-tutorial-examples/) `i` to be a compile-time constant.

However, there seems to be a solution [here](https://www.justsoftwaresolutions.co.uk/cplusplus/getting-tuple-elements-with-runtime-index.html). I will try it.

Reply:
You can use something like:

```c++
template<std::size_t I = 0, typename... Tp>
typename std::enable_if<I == sizeof...(Tp), void>::type
Evaluate(std::tuple<Tp...>& t)
{
  // Nothing to do here.
}

template<std::size_t I = 0, typename... Tp>
typename std::enable_if<I < sizeof...(Tp), void>::type
Evaluate(std::tuple<Tp...>& t)
{
  std::get<I>(t).Evalaute(....);
  Evaluate<I + 1, Tp...>(t);
}


Evaluate(objectives, ...);
```

Let me know if I should clarify anything.

Reply:
Ah got it!

So, when we have 0 objectives, we use the first definition and the second definition otherwise. The second definition recursively calls itself as long as objectives exist.

If I am not wrong, this is SFINAE right? I would love to read up on it! :)

Reply:
Marcus, the `std::enable_if<I == sizeof...(Tp), void>` case is causing the build failure (`-Werror=unused-parameter`) in the CI.

Reply:
You are absolutely right, C++ 17 introduces `index_sequence` which will make the construct easier, but don't think we have C++ 17 support on most of the systems in the wild.

Reply:
`index_sequence` looks really interesting. Especially the ability to do something like [this](https://en.cppreference.com/w/cpp/utility/integer_sequence):
```cpp
template<typename T, T... ints>
void print_sequence(std::integer_sequence<T, ints...> int_seq)
{
    std::cout << "The sequence of size " << int_seq.size() << ": ";
    ((std::cout << ints << ' '),...);
    std::cout << '\n';
}
```
would be great for debugging (with print statements 😛)!